tweaks to DevicesSharding / ReshapeableDevicesSharding

mattjj opened the pull request:
1. rename DevicesSharding -> ReshapeableDevicesSharding
2. fix repr to print device order faithfully
3. respect shape of np.ndarray argument to __init__

yashk2810 commented on tests/array_test.py:
```diff
@@ -730,7 +730,7 @@ def test_devices_sharding_op_sharding_lowering(
 
-    devices_sharding = sharding.DevicesSharding(jax.devices())
+    devices_sharding = sharding.ReshapeableDevicesSharding(jax.devices())
     devices_sharding = devices_sharding.reshape(shape).replicate(axes)
```
Maybe add a test where you don't do this reshape? To test that you are actually respecting the user shape?

In this test, you can get it by doing `mesh.devices` to get a shape that's not 1D

mattjj replied:
Done!

mattjj committed:
tweaks to DevicesSharding

1. rename DevicesSharding -> ReshapeableDevicesSharding
2. fix repr to print device order faithfully
3. respect shape of np.ndarray argument to __init__
